src/Middleware/ZendAuthenticationServiceMiddleware.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+namespace ZfcRbac\Middleware;
+
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Zend\Authentication\AuthenticationService;
+use Zend\Stratigility\MiddlewareInterface;
+
+class ZendAuthenticationServiceMiddleware implements MiddlewareInterface
+{
+    /**
+     * @var AuthenticationService
+     */
+    private $authService;
+
+    /**
+     * @var string
+     */
+    private $attributeName;
+
+    /**
+     * ZendAuthenticationServiceMiddleware constructor.
+     *
+     * @param AuthenticationService $authService
+     */
+    public function __construct(AuthenticationService $authService, $attributeName = 'identity')
+    {
+        $this->authService   = $authService;
+        $this->attributeName = $attributeName;
+    }
+
+    /**
+     * @param ServerRequestInterface $request
+     * @param ResponseInterface      $response
+     * @param callable|null          $next
+     * @return ResponseInterface
+     */
+    public function __invoke(ServerRequestInterface $request, ResponseInterface $response, callable $next = null)
+    {
+        if ($request->getAttribute($this->attributeName) === null && $this->authService->hasIdentity()) {
+            $identity = $this->authService->getIdentity();
+            $request  = $request->withAttribute($this->attributeName, $identity);
```

Comment:
Nor sure - if we delegate that to the auth service this opens up the idea to stack multiple middleware like this one. If none find an identity null is fed into the isgranted method and guest should be assumed.